### ISSUE TITLE
external_source doesn't return last incomplete batch

### ISSUE THREAD

**Issue by JoostvDoorn** [NVIDIA/DALI#5199]
### Version

1.31.0

### Describe the bug.

An `fn.external_source` with a callable doesn't seem to return the last batch. I have tried using `LastBatchPolicy.PARTIAL`, but it seems that the last_batch_policy has no effect on external_source. Is there any way to ensure the last batch is always returned? It seems that this is a bug.

### Minimum reproducible example

```python
import numpy as np
from nvidia.dali import pipeline_def
import nvidia.dali.fn as fn
import nvidia.dali.types as types
from nvidia.dali.plugin.pytorch import DALIGenericIterator, LastBatchPolicy

batch_size = 32

class ExternalInputCallable:
    def __init__(self, batch_size):
        self.batch_size = batch_size
        self.length = 100
        self.full_iterations = self.length // batch_size

    def __call__(self, sample_info):
        sample_idx = sample_info.idx_in_epoch
        if sample_idx >= self.length:
            print("STOPPED", sample_idx)
            # Indicate end of the epoch
            raise StopIteration
        return np.array([1], dtype=np.int32), np.array([sample_idx], dtype=np.int32)

@pipeline_def(batch_size=batch_size, num_threads=1, device_id=0)
def callable_pipeline():
    data, label = fn.external_source(source=ExternalInputCallable(batch_size), num_outputs=2, batch=False,
                                       dtype=[types.INT32, types.INT32])
    return data, label

call_pipe = callable_pipeline()
call_pipe.build()

loader = DALIGenericIterator(
    call_pipe,
    last_batch_policy=LastBatchPolicy.PARTIAL,
    output_map=['data', 'label'],
    auto_reset=True,
)
max_label = -1
for batch in loader:
    max_label = max(batch[0]["label"].flatten().max().item(), max_label)
    print(max_label, batch[0]["label"].shape)
print(max_label)
```

We expect max label to return 99.

### Relevant log output

```shell
31 torch.Size([32, 1])
STOPPED 100
63 torch.Size([32, 1])
95 torch.Size([32, 1])
95
```

### Other/Misc.

```
> pip show nvidia-dali-cuda120
Name: nvidia-dali-cuda120
Version: 1.31.0
Summary: NVIDIA DALI  for CUDA 12.0. Git SHA: 166a2e445992d4a0fca41be32ef897ea0a526a6e
Home-page: https://github.com/NVIDIA/dali
Author: NVIDIA Corporation
Author-email: 
License: Apache License 2.0
```

### Check for duplicates

- [X] I have searched the [open bugs/issues](https://github.com/NVIDIA/DALI/issues) and have found no duplicates for this bug report

**Comment by awolant** [comment]
Hello @JoostvDoorn, thank you for creating the issue and thanks for the repro you provided. 

The behavior you are seeing is not controlled by the `DALIGenericIterator` parameters. That's why you see now difference in the output when you change them. You could remove the iterator layer from your repro you would get the same result:
```python
import numpy as np
from nvidia.dali import pipeline_def
import nvidia.dali.fn as fn
import nvidia.dali.types as types

batch_size = 32

class ExternalInputCallable:
    def __init__(self, batch_size):
        self.batch_size = batch_size
        self.length = 100
        self.full_iterations = self.length // batch_size

    def __call__(self, sample_info):
        sample_idx = sample_info.idx_in_epoch
        if sample_idx >= self.length:
            print("STOPPED", sample_idx)
            # Indicate end of the epoch
            raise StopIteration
        return np.array([1], dtype=np.int32), np.array([sample_idx], dtype=np.int32)

@pipeline_def(batch_size=batch_size, num_threads=1, device_id=0)
def callable_pipeline():
    data, label = fn.external_source(source=ExternalInputCallable(batch_size), num_outputs=2, batch=False,
                                       dtype=[types.INT32, types.INT32])
    return data, label

call_pipe = callable_pipeline()
call_pipe.build()

max_label = -1
for i in range(10):
    batch = call_pipe.run()
    max_label = max(batch[1].as_array().flatten().max().item(), max_label)
    
    print(max_label)
```
Output:
```
31
63
STOPPED 100
95
```


What happens here is controlled by the `pipeline` and `ExternalInputCallable` you created. Note, how it is configured:
```python
self.batch_size = batch_size
self.length = 100
self.full_iterations = self.length // batch_size
```

In your example, with `batch_size == 32` you get `full_iterations == 3`. That is why your pipeline returns only 3 iterations and you loose samples beyond. So to make this work as you intend you need to support it in `ExternalInputCallable`.

Just one more note on this: if you plan to use the parallel external source for your use case, you have to implement batch padding rather than partial batch as it is a requirement to make it work in parallel and is mentioned in 
[Parallel External Source docs](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/data_loading/parallel_external_source.html#Epochs)

Hope that helps

**Comment by JoostvDoorn** [comment]
> In your example, with `batch_size == 32` you get `full_iterations == 3`. That is why your pipeline returns only 3 iterations and you loose samples beyond. So to make this work as you intend you need to support it in `ExternalInputCallable`. With current implementation all DALI facilities see `full_iterations == 3` and try to act accordingly which in conclusion gives the result that we see.

I don't actually use the full_iterations, I use the `self.length` to raise the StopIteration if the length is exceeded. But I see your point that it is documented in the [Parallel External Source docs](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/data_loading/parallel_external_source.html#Epochs) that this is a restrictions, I didn't see this in the specific API docs, so wasn't able to find out this behavior. I suppose what I should do is extend it with a padding, and mask output to be able to return partial batches. It would seem reasonable to support this on the library side since the `sample_info` anyway is produced by the library it should be able to control the partial batch generation.

**Comment by awolant** [comment]
Right, I misread the code, apologies. 

I agree, it would be good to support this. In the meantime I can extend the docs of `source` argument to include this information.

**Comment by mrclovvn** [comment]
Hi~

I've encountered the same issue with the fn.external_source callable not returning the last batch. I've also tried using LastBatchPolicy.PARTIAL, but it seems to have no effect on external_source. Have you found a solution to this problem, or is there an alternative workaround? I look forward to your response.

Best regards.

**Comment by JanuszL** [comment]
Hi @mrclovvn,

As @awolant stated, the behavior is up to the `external_source` itself. So your callable should control this, or in parallel mode, you need to pad samples on your own without the ability to return only partial results. We have an appropriate improvement on our ToDo list.

**Comment by SUW96** [comment]
Hi everyone,

I just ran into this issue. For now, I will implement a Paddington's strategy for my external source, but I am curious if there is already some work done on this topic.
The opportunity to have an external source, that returns individual samples, works in parallel mode and is able to return partial batches would be great.

Best regards

**Comment by JanuszL** [comment]
Hi @SUW96,

Thank you for reaching out. Unfortunately, no work has been done in that direction yet, so padding remains the most reliable strategy for now.

**Comment by eadwu** [comment]
How do I get epoch based resetting? I am raising `StopIteration` but instead it just hangs after prefetching the last batch. I ensure all batches are the full batch size.

```py
class CustomBatchIterator:
    def __init__(self, dataloader):
        self.dataloader = dataloader
        self.iterator = iter(self.dataloader)
        
    def __iter__(self):
        # An iterator must return itself when passed to iter()
        return self
        
    def __next__(self):
        try:
            print("NEXT")
            batch, labels = next(self.iterator)
        except StopIteration:
            # Optional: Reset the iterator automatically for infinite loops
            # self.iterator = iter(self.dataloader)
            # batch = next(self.iterator)
            print("RAISING")
            raise StopIteration
            
        return batch, labels

...

    batch, labels = fn.external_source(
        source=CustomBatchIterator(dataloader), num_outputs=2, # 2-tuple
        batch=True, # pre-batched
        parallel=False, # main-thread feeder,
        cycle="raise", # ??? does nothing with or without
    )
```

Batch counter output
```
326 
NEXT
RAISING
327
328 
```

`DALIGenericIterator` with precomputed `size` works for me since I know it but is not a visually clean solution.

**Comment by JanuszL** [comment]
Hi @eadwu,

Thank you for reaching out.
For epoch-style behavior with external_source, `cycle="raise"` is the right knob, but it only works for resettable iterable sources or generator functions. The source must be something for which `iter(source)` starts a new epoch. In your example, `CustomBatchIterator.__iter__` returns `self` and the inner `self.iterator` is created only once in `__init__`, so after it is exhausted DALI cannot rewind it.

Try making `__iter__` recreate the dataloader iterator:
```
class CustomBatchIterator:
    def __init__(self, dataloader):
      self.dataloader = dataloader
      self.iterator = None

    def __iter__(self):
      self.iterator = iter(self.dataloader)
      return self

    def __next__(self):
      batch, labels = next(self.iterator)
      return batch, labels
```
or pass a generator function, not a generator object:
```
def custom_batches():
  for batch, labels in dataloader:
    yield batch, labels

batch, labels = fn.external_source(
  source=custom_batches,
  num_outputs=2,
  batch=True,
  parallel=False,
  cycle="raise",
)
```

Then the consumer still needs to reset the DALI pipeline/iterator when `StopIteration` is observed. After `pipe.run()` raises StopIteration, call `pipe.reset()` before the next epoch. With `DALIGenericIterator`, use `auto_reset=True` or call `iterator.reset()` explicitly.

One more detail: because DALI prefetches, your source can print `RAISING` before the last already-prefetched batch is returned. That by itself is expected; the `StopIteration` is observed by the pipeline/iterator after the queued batches are drained. The important part is that after that point the source must be rewindable as above.

If that still doesn't work as expected please provide a full, self-contained repro we can run on our end to illustrate the problem you are facing

**Comment by eadwu** [comment]
I am not sure if I am doing something wrong. I don't see any StopIteration getting propagated. I tried both generator/iterator class and iterator/.run approaches

It always stalls once everything is prefetched (936 iterations, never feeds 935/936) so loop hangs
```
Prefetched 936, batch size 64
Consumed 933
prefetched every iteration
Consumed 934
```

Self enclosed with MNIST download

```py
import numpy as np

from nvidia.dali import pipeline_def
import nvidia.dali.fn as fn
import nvidia.dali.types as types

class CustomBatchIterator:
    def __init__(self, dataloader):
      self.dataloader = dataloader
      self.iterator = None

    def __iter__(self):
      self.iterator = iter(self.dataloader)
      return self

    def __next__(self):
      batch, labels = next(self.iterator)
      batch = [np.frombuffer(b, dtype=np.uint8) for b in batch]
      return batch, labels

@pipeline_def(prefetch_queue_depth={ "cpu_size": 2, "gpu_size": 2 })
def pipeline(dataloader):
    def fixed_dl():
        for i, (batch, labels) in enumerate(dataloader):
            print(f"Prefetched {i}, batch size {len(batch)}")
            yield [np.frombuffer(b, dtype=np.uint8) for b in batch], labels
        print("prefetched every iteration")

    jpegs, labels = fn.external_source(
        # source=CustomBatchIterator(dataloader), 
        source=fixed_dl,
        num_outputs=2, # 2-tuple
        batch=True, # pre-batched
        parallel=False, # main-thread feeder,
        cycle="raise",
    )

    images = fn.decoders.image(
        jpegs, device="mixed", hw_decoder_load=0.65,
        output_type=types.DALIImageType.RGB,
    )

    return images, labels


import torch
import torchvision.datasets

def main():
    import io
    def raw_byte_jpg(pil):
       buffer = io.BytesIO()
       pil.save(buffer, format='JPEG')
       return buffer.getvalue()
    dataset = torchvision.datasets.MNIST(root='/tmp/mnist', download=True, transform=raw_byte_jpg)

    # not cleanly batch-able
    print(len(dataset), len(dataset) / 64.0, len(dataset) % 64)
    # drop last partial batch
    dataloader = torch.utils.data.DataLoader(dataset, batch_size=64, drop_last=True)

    from nvidia.dali.plugin.pytorch import DALIGenericIterator
    from nvidia.dali.plugin.base_iterator import LastBatchPolicy
    ds = DALIGenericIterator(
        [pipeline(
            dataloader, 
            batch_size=64, num_threads=4, 
            # device_id=0,
        )],
        output_map=["images", "labels"],
    )

    x = pipeline(
        dataloader, 
        batch_size=64, num_threads=4, 
        # device_id=0,
    )

    for epoch in range(10):
        print(f"Epoch: {epoch+1}")
        # for i, batch in enumerate(ds):
        #     print(f"Consumed {i}")
        i = 0
        while True:
            try:
                batch = x.run()
                print(f"Consumed {i}")
            except StopIteration:
                print("stop received?")
                break
            i = i + 1
        ds.reset()


if __name__ == "__main__":
    main()
```

Version is installed from NVIDIA pypi `nvidia_dali_cuda130-2.1.0.dist-info`
```
    "nvidia-dali-cuda130>=2.1.0",

[tool.uv.sources]
nvidia-dali-cuda130 = { index = "nvidia" }

[[tool.uv.index]]
name = "nvidia"
url = "https://pypi.nvidia.com"
explicit = true
```

**Comment by JanuszL** [comment]
Hi @eadwu,

I can confirm that `prefetch_queue_depth={ "cpu_size": 2, "gpu_size": 2 }` is the issue. If you just set `prefetch_queue_depth=2` it should work correctly. In the meantime we will investigate why separated prefetch queues lead to hang.

**Comment by JanuszL** [comment]
https://github.com/NVIDIA/DALI/pull/6397 has been merged. Please check the next nightly build after today, and is should be available in 2.3 release.
